Mathlib/MeasureTheory/Function/FactorsThrough.lean

```diff
@@ -3,47 +3,78 @@ Copyright (c) 2025 Etienne Marion. All rights reserved.
+import Mathlib.MeasureTheory.Constructions.Polish.StronglyMeasurable
 import Mathlib.Probability.Process.Filtration
 
 /-!
 # Factorization of a map from measurability
 
 Consider `f : X → Y` and `g : X → Z` and assume that `g` is measurable with respect to the pullback
-along `f`. Then `g` factors though `f`, which means that there exists `h : Y → Z` such that
-`g = h ∘ f`.
+along `f`. Then `g` factors through `f`, which means that (if `Z` is nonempty)
+there exists `h : Y → Z` such that `g = h ∘ f`.
 
-## TODO
-
-* Under certain assumptions, the factorization map `h` is measurable. This is the content of the
-  [Doob-Dynkin lemma](https://en.wikipedia.org/wiki/Doob–Dynkin_lemma).
+If `Z` is completely metrizable, the factorization map `h` can be taken to be measurable.
+This is the content of the [Doob-Dynkin lemma](https://en.wikipedia.org/wiki/Doob–Dynkin_lemma):
+see `exists_eq_measurable_comp`.
 -/
 
 namespace MeasureTheory
 
-open Filtration Set TopologicalSpace
+open Filter Filtration Set TopologicalSpace
+
+open scoped Topology
 
 variable {X Y Z : Type*} [mY : MeasurableSpace Y] {f : X → Y} {g : X → Z}
 
+section FactorsThrough
+
 /-- If a function `g` is measurable with respect to the pullback along some function `f`,
 then to prove `g x = g y` it is enough to prove `f x = f y`. -/
 theorem _root_.Measurable.factorsThrough [MeasurableSpace Z] [MeasurableSingletonClass Z]
     (hg : Measurable[mY.comap f] g) : g.FactorsThrough f := by
   refine fun x₁ x₂ h ↦ eq_of_mem_singleton ?_
   obtain ⟨s, -, hs⟩ := hg (measurableSet_singleton (g x₂))
-  rw [← mem_preimage, ← hs, mem_preimage, h, ← mem_preimage, hs, mem_preimage, mem_singleton_iff]
+  rw [← mem_preimage, ← hs, mem_preimage, h, ← mem_preimage, hs]
+  rfl
 
 /-- If a function `g` is strongly measurable with respect to the pullback along some function `f`,
 then to prove `g x = g y` it is enough to prove `f x = f y`.
 
-TODO: under certain assumptions, the factorization map `h` is measurable. This is the content of the
-[Doob-Dynkin lemma](https://en.wikipedia.org/wiki/Doob–Dynkin_lemma). -/
+If `Z` is not empty there exists `h : Y → Z` such that `g = h ∘ f`.
+If `Z` is also completely metrizable, the factorization map `h` can be taken to be measurable
+(see `exists_eq_measurable_comp`). -/
 theorem StronglyMeasurable.factorsThrough [TopologicalSpace Z]
     [PseudoMetrizableSpace Z] [T1Space Z] (hg : StronglyMeasurable[mY.comap f] g) :
     g.FactorsThrough f := by
   borelize Z
   exact hg.measurable.factorsThrough
 
+variable {ι : Type*} [TopologicalSpace Z] [IsCompletelyMetrizableSpace Z]
+  [Countable ι] {l : Filter ι} [l.IsCountablyGenerated] {f : ι → X → Z}
+
+/-- If a function `g` is strongly measurable with respect to the pullback along some function `f`,
+then there exists some measurable function `h : Y → Z` such that `g = h ∘ f`. -/
+theorem StronglyMeasurable.exists_eq_measurable_comp [Nonempty Z] {f : X → Y} {g : X → Z}
```

Comment:
What about giving also a version where you assume that `X` is a measurable space and `f` is measurable?